devito/at_controller.py

```diff
@@ -0,0 +1,223 @@
+        model_descr_str = "%s %d %d %s %s" % (f_name, time_order, spc_border,
+                                              str(shape).replace(" ", ''), cb_dims)
+
+        with open(final_report_path, 'r') as f:
+            for line in f.readlines():
+
+                if model_descr_str in line:
+                    blocks_str = line.split(' ')[5]
+                    block_split = blocks_str[1:len(blocks_str) - 2].split(',')
+
+                    return [int(block) if block != "None" else None for block in block_split]
+
+        return None  # returns none if no matching block size was found
+
+    @staticmethod
+    def get_optimal_block_size(shape, load_c):
+        """
+        Gets optimal block size based on architecture
+        :param shape: list - shape of the data buffer
+        :param load_c: int - load count
+        :return: optimal block size
+        """
+        cache_s = int(cpuinfo.get_cpu_info()['l2_cache_size'].split(' ')[0])  # cache size
+        core_c = cpuinfo.get_cpu_info()['count']  # number of cores
+
+        # assuming no prefetching, square block will give the most cache reuse.
+        # We then take the cache/core divide by the size of the inner most dimension in which we do not block.
+        #  This gives us the X*Y block space, of which we take the square root to get the size of our blocks.
+        # ((C size / cores) / (4 * length inner most * kernel loads)
+        optimal_b_size = math.sqrt(((1000 * cache_s) / core_c) / (4 * shape[len(shape) - 1] * load_c))
+        return int(round(optimal_b_size))  # rounds to the nearest integer
+
+    def auto_tune_blocks(self, minimum=5, maximum=20):
+        """
+        Auto tunes block sizes. Times all block size combinations withing given range and writes it into report
+        :param minimum: int (optional) - minimum value for auto tuning range. Default 5
+        :param maximum: int (optional) - maximum value for auto tuning range. Default 20
+        :raises ValueError: if  minimum is >= maximum
+        :raises EnvironmentError: If cache blocking is not set inside the Operator object
+        """
+        if minimum >= maximum:
+            raise ValueError("Invalid parameters. Minimum tune range has to be less than maximum")
+
+        if not self.op.propagator.cache_blocking:
+            raise EnvironmentError("Invalid parameters. Cache_blocking is set to False")
+
+        logger.info("Starting auto tuning of block sizes using brute force")
+
+        if maximum - minimum > 2:  # No point to estimate if diff is below 2.
+            self._estimate_run_time(minimum, maximum)
+
+        times = []  # list where times and block sizes will be kept
+        block_list = []  # used to make sure we do not test the same block sizes
+        org_block_sizes = self.op.propagator.block_sizes
+        blocks = list(self.op.propagator.block_sizes)
+
+        for x in range(minimum, maximum):
+            blocks[0] = x if org_block_sizes[0] else None
```

Comment:
Forget it, I think I understand.

Indexing into lists directly as you do here can be difficult to understand sometimes. Maybe use a ``namedtuple`` ?